### PR TITLE
PaydirektEditViewController layout

### PR DESCRIPTION
**Pull request by mobileapps23**
- programmatic layout for PaydirektEditViewController

### CODE REVIEW

File: Snabble/UI/de.lproj/SnabbleLocalizable.strings [hunk 1]
```diff
@@ -104,6 +104,8 @@
 
 "Snabble.OK" = "OK";
 
+"Snabble.paydirekt.title" = "paydirekt";
```
Comment (mrgrauel) [inline Snabble/UI/de.lproj/SnabbleLocalizable.strings]: We use phrase.com for translations. 8caa630bdb1e0966973630cba7c96aeaa5ccd0ca
If you find something like that let me know.

**Review by mrgrauel — CHANGES_REQUESTED** [review]

File: Snabble/UI/PaymentMethods/PaydirektEditViewController.swift [hunk 2]
```diff
@@ -56,42 +56,109 @@ public final class PaydirektEditViewController: UIViewController {
         redirectUrlAfterFailure: RedirectStatus.failure.url
     )
 
-    public init(_ detail: PaymentMethodDetail?, _ analyticsDelegate: AnalyticsDelegate?) {
+    public init(_ detail: PaymentMethodDetail?, with analyticsDelegate: AnalyticsDelegate?) {
```
[the remark this answers is not in the document]
Comment (mobileapps23) [reply Snabble/UI/PaymentMethods/PaydirektEditViewController.swift:59]: Then should be also changed init methods for `TeleCashCreditCardEditViewController`, `PayoneCreditCardEditViewController`, `SepaEditViewController`, `PayoneCreditCardEditViewController`, `DatatransAliasViewController`, etc.
And logic for creation entry points:
 at `MethodRegistry` -> 
`createEntry(method: RawPaymentMethod, _ projectId: Identifier<Project>, _ analyticsDelegate: AnalyticsDelegate?) -> UIViewController?`
at `DatatransFactory` enum, etc.
Do we need this kind of refactoring now? I can make it as separate pull request, if yes.

File: Snabble/UI/PaymentMethods/PaydirektEditViewController.swift [hunk 3]
```diff
@@ -35,11 +35,11 @@ private enum RedirectStatus: String {
...
-    @IBOutlet private var displayLabel: UILabel!
-    @IBOutlet private var openButton: UIButton!
-    @IBOutlet private var deleteButton: UIButton!
+    private weak var webViewWrapper: UIView?
+    private weak var displayView: UIView?
+    private weak var displayLabel: UILabel?
+    private weak var openButton: UIButton?
+    private weak var deleteButton: UIButton?
 
     private var webView: WKWebView?
```
[the remark this answers is not in the document]
Comment (mobileapps23) [reply Snabble/UI/PaymentMethods/PaydirektEditViewController.swift]: Fixed

File: Snabble/UI/PaymentMethods/PaydirektEditViewController.swift [hunk 4]
```diff
@@ -56,42 +56,109 @@ public final class PaydirektEditViewController: UIViewController {
...
     required init?(coder aDecoder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }
 
+    override public func loadView() {
+        super.loadView()
+
+        let view = UIView(frame: UIScreen.main.bounds)
+        view.backgroundColor = .systemBackground
+        self.view = view
```
[the remark this answers is not in the document]
Comment (mobileapps23) [reply Snabble/UI/PaymentMethods/PaydirektEditViewController.swift]: Fixed

File: Snabble/UI/PaymentMethods/PaydirektEditViewController.swift [hunk 5]
```diff
@@ -56,42 +56,109 @@ public final class PaydirektEditViewController: UIViewController {
...
-        self.displayLabel.text = L10n.Snabble.Paydirekt.savedAuthorization
-
-        self.deleteButton.makeSnabbleButton()
-        self.deleteButton.setTitle(L10n.Snabble.Paydirekt.deleteAuthorization, for: .normal)
-
-        self.openButton.setTitle(L10n.Snabble.Paydirekt.gotoWebsite, for: .normal)
+        self.title = L10n.Snabble.Paydirekt.title
+        if let wrapper = self.webViewWrapper {
+            self.webView = self.addWebView(to: wrapper)
+        }
```
[the remark this answers is not in the document]
Comment (mobileapps23) [reply Snabble/UI/PaymentMethods/PaydirektEditViewController.swift]: Fixed

File: Snabble/UI/PaymentMethods/PaydirektEditViewController.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrgrauel) [reply Snabble/UI/PaymentMethods/PaydirektEditViewController.swift:59]: No, you are correct that's out of scope.

**Review by mrgrauel — CHANGES_REQUESTED** [review]
[Screen Shot 2022-04-21 at 15 54 19](https://user-images.githubusercontent.com/148758/164473527-a7d807dc-8547-4c5f-ae17-f0ee9895debf.png)

I see an empty screen when I try to add Paydirekt for Tegut (Environment: Testing)

**Comment by mobileapps23** [comment]
When try to add Paydirekt for Tegut (Environment: Testing) there is an empty screen because link for paydirekt is `nil` in `Metadata`:
line 163 - `let authUrl = Snabble.metadata.links.paydirektCustomerAuthorization?.href,`
output for `links` is -
`MetadataLinks(
clientOrders: Optional(SnabbleSDK.Link(href: "/clients/{clientID}/orders")), 
appUser: SnabbleSDK.Link(href: "/apps/users/{appUserID}"), 
appUserOrders: SnabbleSDK.Link(href: "/apps/users/{appUserID}/orders"), 
consents: Optional(SnabbleSDK.Link(href: "/apps/users/{appUserID}/consents")), 
paydirektCustomerAuthorization: nil, 
createAppUser: SnabbleSDK.Link(href: "/apps/snabble-app-quae3M/users"), 
_self: SnabbleSDK.Link(href: "/metadata/app/snabble-app-quae3M/ios/2.9.1"))`

and there is no handling for this case(and also for every failure case at `func startAuthorization()`).
It's better to add an alert, and need to know title and message for it.

Another question is about line 164 - `let project = Snabble.projects.first`
Is it correct way here to define current project without its id? Because it always will give an 'Aldi', as it is the first one in the array. I guess, it should be changed to` Snabble.project(for projectId: Identifier<Project>)`, and `viewController` should be initialized with `projectId` parameter.

**Comment by mrgrauel** [comment]
It's all correct 👍 

Do you want to refactor it as you mentioned? For me the network layer shouldn't need a `project` instance, but that's something we have to do in the future.
